src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs

```diff
@@ -30,21 +48,56 @@ public async Task<Checksum> GetChecksumAsync(CancellationToken cancellationToken
             return collection.Checksum;
         }
 
-        private async Task<SolutionStateChecksums> ComputeChecksumsAsync(CancellationToken cancellationToken)
+        /// <param name="projectId">If specified, the checksum will only contain information about the 
+        /// provided project (and any projects it depends on)</param>
+        public async Task<Checksum> GetChecksumAsync(ProjectId? projectId, CancellationToken cancellationToken)
+        {
+            if (projectId == null)
+                return await GetChecksumAsync(cancellationToken).ConfigureAwait(false);
+
+            ValueSource<SolutionStateChecksums>? lazyChecksum;
+            lock (_lazyProjectChecksums)
+            {
+                if (!_lazyProjectChecksums.TryGetValue(projectId, out lazyChecksum))
+                {
+                    lazyChecksum = CreateLazyChecksum(projectId);
+                    _lazyProjectChecksums.Add(projectId, lazyChecksum);
+                }
+            }
+
+            var collection = await lazyChecksum.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            return collection.Checksum;
+
+            // Extracted as a local function to prevent delegate allocations when not needed.
```

Comment:
>allocations [](http://example.com/codeflow?start=65&length=11)

Not sure why this is needed. The closure would only be allocated when the containing block is entered, which is exactly when the locla function is called.

Reply:
oh?  i thought it was allocated at the toplevel method boundary.

Reply:
Yeah, you're right - I missed that projectId is a parameter. You could write it as 

```C#
{
  var capturedProjectId = projectId;
  lazyChecksum = new AsyncLazy<SolutionStateChecksums>(
                    c => ComputeChecksumsAsync(capturedProjectId, c), cacheResult: true)
}
```

Reply:
Ah. Cute. I'll do that!